### PR TITLE
feat: allow annotations on enums

### PR DESCRIPTION
**Pull request by nicdard**
Partially address https://github.com/flix/flix/issues/2831

I just encoded enums to start looking into the issue and get some initial feedback, in a following PR I will do the same also for classes and type aliases.

This PR builds on top of: https://github.com/flix/flix/pull/3289

### CODE REVIEW

**Comment by magnus-madsen** [comment]
@mlutze Can you take an initial look? Thanks!

**Review by mlutze — CHANGES_REQUESTED** [review]
Looks good! Just a couple little things.

File: main/src/ca/uwaterloo/flix/language/phase/OccurrenceAnalyzer.scala [hunk 1]
```diff
@@ -36,6 +36,7 @@ object OccurrenceAnalyzer {
    */
   def run(root: LiftedAst.Root)(implicit flix: Flix): Validation[OccurrenceAst.Root, CompilationMessage] = flix.phase("OccurrenceAnalyzer") {
 
+    // Visit every definition in the program in parallel and transform to type 'OccurrenceAst.Def'
```
[the remark this answers is not in the document]
Comment (nicdard) [reply main/src/ca/uwaterloo/flix/language/phase/OccurrenceAnalyzer.scala:39]: Oh sorry

File: main/src/ca/uwaterloo/flix/language/phase/Simplifier.scala [hunk 2]
```diff
@@ -788,11 +788,12 @@ object Simplifier {
     //
     val defns = root.defs.map { case (k, v) => k -> visitDef(v) }
     val enums = root.enums.map {
-      case (k, TypedAst.Enum(_, mod, sym, _, _, cases0, enumType, _, loc)) =>
+      case (k, TypedAst.Enum(_, ann, mod, sym, _, _, cases0, enumType, _, loc)) =>
         val cases = cases0 map {
           case (tag, TypedAst.Case(enumSym, tagName, tagType, _, tagLoc)) => tag -> SimplifiedAst.Case(enumSym, tagName, tagType, tagLoc)
         }
-        k -> SimplifiedAst.Enum(mod, sym, cases, enumType, loc)
+        val sAnn = if (ann.isEmpty) Ast.Annotations.Empty else Ast.Annotations(ann.map(a => a.name))
```
[the remark this answers is not in the document]
Comment (nicdard) [reply main/src/ca/uwaterloo/flix/language/phase/Simplifier.scala:795]: Yes I was also wondering about this, so I will change it, should I also change it in the other point of the code for consistency?

[the remark this answers is not in the document]
Comment (mlutze) [reply main/src/ca/uwaterloo/flix/language/phase/Simplifier.scala]: Yes please :)

**Review by mlutze — APPROVED** [review]

File: main/src/ca/uwaterloo/flix/language/phase/Weeder.scala [hunk 3]
```diff
@@ -223,13 +223,14 @@ object Weeder {
...
+    case ParsedAst.Declaration.Enum(doc0, ann, mods, sp1, ident, tparams0, derives, cases, sp2) =>
       val doc = visitDoc(doc0)
+      val annVal = visitAnnotations(ann)
       val modVal = visitModifiers(mods, legalModifiers = Set(Ast.Modifier.Public))
       val tparamsVal = visitTypeParams(tparams0)
 
-      flatMapN(modVal, tparamsVal) {
-        case (mod, tparams) =>
+      flatMapN(modVal, annVal, tparamsVal) {
+        case (mod, ann, tparams) =>
```
[the remark this answers is not in the document]
Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/phase/Weeder.scala:233]: i.e. flatMapN(annVal, modVal, ...)

**Review by magnus-madsen — COMMENTED** [review]
Looks good. I added one small comment.

One thing that is missing: We should use these annotations in `Documentor`. See here how it is done for defs: https://github.com/flix/flix/blob/master/main/src/ca/uwaterloo/flix/language/phase/Documentor.scala#L116

We should also use the annotations in LSP, e.g. if something is marked deprecated we should show it. That happens in `CodeHinter`.

File: main/src/ca/uwaterloo/flix/language/phase/Weeder.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (nicdard) [reply main/src/ca/uwaterloo/flix/language/phase/Weeder.scala]: Thanks!

**Comment by nicdard** [comment]
> Excellent! Glad to have this merged in :)

I will also port the suggestions to https://github.com/flix/flix/pull/3294 as they are similar :)

**Pull request merged**